ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,969 @@
+    !(LedgerState blk ValuesMK)
+    !IsReachable -- ^ Is this a new fork? i.e. should the current mempool tip became
+                 -- unreachable?
+  deriving (Generic1)
+  deriving (Rank2.Functor, Rank2.Foldable, Rank2.Traversable, CommandNames)
+
+instance CommandNames (Command blk) where
+  cmdName (Action action) = cmdName action
+  cmdName (Event event)   = cmdName event
+
+  cmdNames :: forall r. Proxy (Command blk r) -> [String]
+  cmdNames _ = cmdNames (Proxy @(Action blk r))
+            ++ cmdNames (Proxy @(Event blk r))
+
+-- | Wether or not this test must be atomic.
+--
+-- The reason behind this data type is that 'TryAddTxs' is on its nature prone
+-- to race-conditions. And that is OK with us. For example take the following
+-- sequence of commands:
+--
+-- @@@
+--  TryAddTxs [Tx1, Tx2] || GetSnapshot
+-- @@@
+--
+-- If we happen to hit the following interleaving:
+--
+-- @@@
+--  AddTx Tx1; GetSnapshot; AddTx Tx2
+-- @@@
+--
+-- the model will never be able to reproduce the result of the snapshot.
+--
+-- So in order to do a meaningful testing, what we do is:
+--
+-- 1. Run a sequential test of actions ensuring that the responses of the model
+--    and SUT match on 'GetSnaphsot'. This provides us with assurance that the
+--    model works as expected on single-threaded/sequential scenarios.
+--
+-- 2. Run a parallel test where 'TryAddTxs' is unitary (i.e. use the 'Atomic'
+--    modifier) ensuring that the responses of the model and SUT match on
+--    'GetSnaphsot'. This ensures that there are no race conditions on this
+--    case, or rephrased, that the operations on the mempool remain atomic even
+--    if executed on separate threads.
+--
+-- 3. Run a parallel test where 'TryAddTxs' is not unitary (using the
+--    'NonAtomic' modifier) and **NOT** checking the responses of the model
+--    versus the SUT. This ensures that there are no deadlocks and no
+--    errors/exceptions thrown when running in parallel.
+--
+-- We believe that these test cover all the interesting cases and provide enough
+-- assurance on the implementation of the Mempool.
+data MakeAtomic = Atomic | NonAtomic | DontCare
+
+generator ::
+  ( Arbitrary (LedgerState blk ValuesMK)
+  , GetTip (LedgerState blk ValuesMK)
+  , Eq (LedgerState blk ValuesMK)
+  , UnTick blk
+  )
+  => MakeAtomic
+  -> (Int -> LedgerState blk ValuesMK -> Gen [GenTx blk])
+     -- ^ Transaction generator based on an state
+  -> Model blk Symbolic
+  -> Maybe (Gen (Command blk Symbolic))
+generator ma gTxs model =
+   Just $
+    frequency
+      [ (100,
+          case ma of
+            Atomic -> do
+              Action . TryAddTxs <$> (gTxs 1 . unTick $ modelMempoolIntermediateState)
+            _ -> do
+              n <- getPositive <$> arbitrary
+              Action . TryAddTxs <$> (gTxs n . unTick $ modelMempoolIntermediateState)
+        )
+      , (10, pure $ Action SyncLedger)
+      , (10, do
+            ls <- arbitrary `suchThat`
```

Comment:
I think this needs to be removed since the shrinker does not guarantee this such that (and the condition is just way too complex). I'd rather skip the change ledger action in the transition and semantics if the tip does not change.